Multitenancy

rebecca-caktus opened the pull request:
Basic integration of multitenancy (using https://github.com/theirc/rapidsms-multitenancy).

Some things which are still needed:
- [ ] More tests and better test structure.
- [ ] In the admin, validate that an object's foreign keys refer to the same tenant
- [ ] Move generic things to rapidsms-multitenancy
- [ ] Remove all dependency on `decisiontree.multitenancy` from `decisiontree`
- [ ] Run tests with and without multitenancy enabled.

Commit:
Add {% tenancy_url %} tag & tests

Commit:
Basic decisiontree multitenancy package

Commit:
First pass at decision tree multitenancy admin

Commit:
Filter views to only tenant-related models

Commit:
Add multitenancy to test requirements for now

Commit:
Dry out multitenancy admin code.

Commit:
Limit foreign key choices to objects for tenants the user manages.

Commit:
TenantMixin -> TenantViewMixin

Commit:
Disallow mixing args and kwargs in tenancy_url tag

Commit:
Update survey list template HTML

Commit:
Tenancy url resolution in create/edit/delete views.

Commit:
Account for tenancy in model forms

Commit:
Use generic linked model field name in TenantLink models

Commit:
Allow superusers to view objects with no tenant in admin

Commit:
Fix Flake8 mistakes

Commit:
Update survey create/update HTML

Commit:
Consolidate and update templates; standardize template names

Commit:
Remove unused template tags and templates

Commit:
Make tenancy_reverse a generic utility

Commit:
Squash multitenancy migrations

Commit:
More extensible way of registering multitenancy models.

Commit:
Fix references in model __str__() methods

Commit:
Create single TenantLink base model

Commit:
Update template for entries list

Commit:
Remove unused static images

Commit:
Correct block name for extra stylesheets; use CSS for form links

Commit:
Add signals to handle tenant link creation

For models with derivable tenant links.

Commit:
Don't display tenant link inline for models with derived tenant relationships.

Commit:
Get tests running

vkurup commented on decisiontree/forms/fields.py:
```diff
@@ -0,0 +1,28 @@
+    widget = TagWidget
+
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault('help_text', 'Tags with spaces must be quoted, for '
```
This could be i18n'd

rebecca-caktus replied:
Added #21 to cover this throughout the app.

vkurup commented on decisiontree/forms/forms.py:
```diff
@@ -0,0 +1,114 @@
+        # self.fields['answer'].queryset = answers.distinct()
+        self.fields['tag'].queryset = tags
+        # self.fields['analysis'].label = 'Calculator'
+        # self.fields['tag'].label = 'Calculator'
```
I couldn't tell whether it's worth keeping these commented lines around, or not.

rebecca-caktus replied:
I suggest that we keep them around for a little while. The answer search form actually isn't hooked up properly (it renders but filling it out has no effect), but I haven't analyzed it fully to see how it could/should work.

vkurup commented on decisiontree/multitenancy/admin.py:
```diff
@@ -0,0 +1,103 @@
+        })
+
+    def get_inline_instances(self, request, obj=None):
+        """Add an inline fo the object's tenant link."""
```
typo.

vkurup commented on decisiontree/templatetags/tree_tags.py:
```diff
@@ -43,16 +45,19 @@ def mode(values):
-    return {"tree": tree}
+@register.simple_tag(takes_context=True)
+@register.assignment_tag(takes_context=True, name='assign_tenancy_url')
+def tenancy_url(context, url_name, *args, **kwargs):
```
Such a good idea!

vkurup commented:
I've reviewed this and am floored by how clean this PR is. (I wish I had thought of doing a lot of this during my initial attempts at multitenancy in the other apps I've worked on). There are some strings that could be i18n'd, but that can be left to a separate PR down the road.

Amazing work! :star2:

Commit:
Fix typo in comment

Commit:
Fix error if passing a list of args to tenancy_reverse

Commit:
Create is_multitenant_model utility

And more reliable traversal of the Django field hierarchy to retrieve the
link model class.

Commit:
Flake8 fix

Newest version of flake8 doesn't allow line to begin with a binary operator.

ghost commented:
Thanks for the kind words, Vinod!
